### PR TITLE
Zoom controls

### PR DESCRIPTION
**Pull request by NicoRoberts**

### CODE REVIEW

File: DronelinkDJI/DJIAdapters.swift [hunk 1]
```diff
@@ -461,6 +472,30 @@ public struct DJICameraStateAdapter: CameraStateAdapter {
     public var focusMode: DronelinkCore.Kernel.CameraFocusMode { return focusModeValue?.kernelValue ?? .unknown }
     public var meteringMode: DronelinkCore.Kernel.CameraMeteringMode { return meteringModeValue?.kernelValue ?? .unknown }
     public var aspectRatio: Kernel.CameraPhotoAspectRatio { (mode == .photo ? photoAspectRatioValue?.kernelValue : nil) ?? ._16x9 }
+    public var opticalZoomSpecResolved: [String : Int] {
```
Comment (NicoRoberts) [inline DronelinkDJI/DJIAdapters.swift:475]: having this named "opticalZoomSpec" clobbers the instance field so I renamed to have "resolved" at the end

File: DronelinkDJI/DJIDroneSession+CameraCommand.swift [hunk 2]
```diff
@@ -198,6 +198,47 @@ extension DJIDroneSession {
             return nil
         }
         
+        if let command = cameraCommand as? Kernel.ZoomCameraCommand {
+            let zoomMax = Double(state.opticalZoomSpecResolved[Kernel.CameraZoomSpec.max.rawValue] ?? 0)
+            let zoomMin = Double(state.opticalZoomSpecResolved[Kernel.CameraZoomSpec.min.rawValue] ?? 0)
+            let zoomStep = (state.opticalZoomSpecResolved[Kernel.CameraZoomSpec.min.rawValue] == nil
+                            || state.opticalZoomSpecResolved[Kernel.CameraZoomSpec.min.rawValue] == 0) ? 1 : Double(state.opticalZoomSpecResolved[Kernel.CameraZoomSpec.min.rawValue] ?? 0)
```
Comment (NicoRoberts) [inline DronelinkDJI/DJIDroneSession+CameraCommand.swift:205]: These statements are checking for null values as a guard but also sets zoom step to 1 to protect against division by zero

File: DronelinkDJI/DJIAdapters.swift [hunk 3]
```diff
@@ -326,6 +326,10 @@ public struct DJICameraFile : CameraFile {
 }
 
 public struct DJICameraStateAdapter: CameraStateAdapter {
+
+    private let opticalZoomSpec: DJICameraOpticalZoomSpec?
```
Comment (jim-mcandrew) [inline DronelinkDJI/DJIAdapters.swift:330]: I would probably move this down by opticalZoomValue

File: DronelinkDJI/DJIDroneSession+CameraCommand.swift [hunk 4]
```diff
@@ -198,6 +198,24 @@ extension DJIDroneSession {
...
```
Comment (jim-mcandrew) [inline DronelinkDJI/DJIDroneSession+CameraCommand.swift:205]: Do we understand how DJI goes from optical to digital yet?

File: DronelinkDJI/DJIAdapters.swift [hunk 5]
```diff
@@ -461,6 +470,42 @@ public struct DJICameraStateAdapter: CameraStateAdapter {
...
+        let maxOptical = hybridZoomSpec.maxOpticalFocalLength
+        let step = hybridZoomSpec.focalLengthStep
+        if let error = validateZoomSpecArguments(currentZoom: zoomValue, min: min, max: max, maxOptical: maxOptical, step: step) {
+            NSLog("Illegal zoomSpec arguments: \(error)")
+            return nil
+        }
+        return Kernel.PercentZoomSpec(currentZoom: zoomValue, min: min, max: max, maxOptical: maxOptical, step: step)
+    }
+    
+    private func validateZoomSpecArguments(currentZoom: Double, min: UInt, max: UInt, maxOptical: UInt, step: UInt) -> String? {
```
Comment (NicoRoberts) [inline DronelinkDJI/DJIAdapters.swift:489]: move this to PercentZoomSpec and throw exception

Comment (NicoRoberts) [reply DronelinkDJI/DJIAdapters.swift:489]: Look at missionExecutor init?

File: DronelinkDJI/DJIAdapters.swift [hunk 6]
```diff
@@ -461,6 +470,42 @@ public struct DJICameraStateAdapter: CameraStateAdapter {
...
+            min >= max || (max - min) < step ||
+            maxOptical < min || maxOptical > max ||
+            currentZoom < Double(min) || currentZoom > Double(max) {
+            return String("Arguments produce an invalid zoom spec. Arguments must be positive and non-zero, min must be less than max, maxOptical must be between min and max inclusive, and step must be less than max - min.")
+        }
+        return nil
+    }
+    
+    public var isPercentZoomSupported: Bool {
+        guard let camera = camera else {
```
Comment (NicoRoberts) [inline DronelinkDJI/DJIAdapters.swift:500]: take advantage of optional type in swift

File: DronelinkDJI/DJIAdapters.swift [hunk 7]
```diff
@@ -461,6 +470,42 @@ public struct DJICameraStateAdapter: CameraStateAdapter {
...
+            return false
+        }
+        //Only support hybrid zoom
+        return camera.isHybridZoomSupported()
+    }
+    public var isRatioZoomSupported: Bool {
+        return false
```
Comment (NicoRoberts) [inline DronelinkDJI/DJIAdapters.swift:507]: remove "return" and make it 1 line.

File: DronelinkDJI/DJIDroneSession+CameraCommand.swift [hunk 8]
```diff
@@ -198,6 +198,25 @@ extension DJIDroneSession {
             return nil
         }
         
+        if let command = cameraCommand as? Kernel.ZoomPercentCameraCommand {
+            guard let zoomSpec = state.zoomSpec else {
+                return "MissionDisengageReason.command.type.unsupported".localized
+            }
+            let zoomMax = Double(zoomSpec.max)
```
Comment (NicoRoberts) [inline DronelinkDJI/DJIDroneSession+CameraCommand.swift:205]: change all spec values to doubles

File: DronelinkDJI/DJIAdapters.swift [hunk 9]
```diff
@@ -461,6 +470,42 @@ public struct DJICameraStateAdapter: CameraStateAdapter {
...
+        guard isPercentZoomSupported, let hybridZoomSpec = hybridZoomSpec, let zoomValue = zoomValue else {
+            return nil
+        }
+        
+        let min = hybridZoomSpec.minHybridFocalLength
+        let max = hybridZoomSpec.maxHybridFocalLength
+        let maxOptical = hybridZoomSpec.maxOpticalFocalLength
+        let step = hybridZoomSpec.focalLengthStep
+        if let error = validateZoomSpecArguments(currentZoom: zoomValue, min: min, max: max, maxOptical: maxOptical, step: step) {
+            NSLog("Illegal zoomSpec arguments: \(error)")
```
Comment (NicoRoberts) [inline DronelinkDJI/DJIAdapters.swift:483]: Global search for NSLog and replace with os_log

File: DronelinkDJI/DJIDroneSession.swift [hunk 10]
```diff
@@ -489,6 +491,37 @@ public class DJIDroneSession: NSObject {
...
+            //Additionally we have to get the hybrid zoom spec from the camera because at the time of writing this code (June 2023), casting newValue to DJICameraHybridZoomSpec does not work.
+            guard let camera = self?.drone.camera(channel: 0) as? DJICamera else {
+                self?._hybridZoomSpec = nil
+                return
+            }
+            
+            if camera.isHybridZoomSupported() {
+                camera.getHybridZoomSpec { (hybridZoomSpec: DJICameraHybridZoomSpec, error: Error?) in
+                    if (error != nil) {
+                        NSLog("Error getting DJICameraHybridZoomSpec: \(error)")
```
Comment (NicoRoberts) [inline DronelinkDJI/DJIDroneSession.swift:515]: change to os_log

**Pull request merged**